src/linkage.md

```diff
@@ -269,6 +272,42 @@ binary link:
 
 Passing `rlib`s directly into your foreign linker is currently unsupported.
 
+### Prohibited linkage and foreign unwinding
+
+r[link.foreign-code.prohibited]
+
+Undfined behavior may be caused by foreign code unwinding into a Rust crate
+with these characteristics:
+
+* The foreign unwind enters Rust via a function or function pointer declared
+  with an ABI that permits unwinding, that is, `"Rust"` (the default ABI) or
+  any `-unwind` ABI
+* The Rust crate containing the `-unwind` ABI declaration was compiled with
+  `panic=unwind`
+* The final binary is linked with [the `panic=abort` runtime][panic-runtime]
```

Comment:
You changed the meaning of the text by making this about the final binary. I am not sure whether this is correct. I'd prefer if you could just restore the old wording, which was reviewed by a bunch of people, and leave further improvements to a later PR so that the diff over the previous consensus can be more easily reviewed.

Reply:
In fact I am fairly sure your new wording is wrong. If the unwinding comes from an external crate, enters through a -Cpanic=unwind crate, and then propagates into a -Cpanic=abort that is UB, even if the binary is -Cpanic=unwind.

Reply:
We certainly talked about the original text a lot, but my impression was that you still considered it confusing. I started the change because it seemed related to the entirely-new "foreign linking" section, but seemed even more confusing in its original form when juxtaposed with the new section. 

I believe the rule has always been about whether or not the "abort" runtime is linked, and since only one panic runtime can be linked in a final Rust binary, I don't understand how the introduction of the word "final" changes this. I can remove the word "final", though. 

A panic=abort crate will have shims for aborting the process in the presence of unwinding at `extern` boundaries. If there's no extern boundary, then the two crates are compiled as source code so they can't have different panic modes in the first place.

Reply:
> If there's no extern boundary, then the two crates are compiled as source code so they can't have different panic modes in the first place.

That's not right. You can link together different crates with different flags.

Reply:
The old text was:
```
No crate may be linked with [the `panic=abort` runtime][panic-runtime] if it has
both of the following characteristics:
* It contains a call to an `-unwind` foreign function or function pointer
* It was compiled with `panic=unwind`
```
Actually looking at this I think there is a mistake here -- the problematic case is linking a crate with the "unwind" runtime, but compiling it with `panic=abort`, isn't it?

Reply:
```suggestion
* The foreign unwind enters Rust via a function or function pointer declared
  with an ABI that permits unwinding, that is, `"Rust"` (the default ABI) or
  any `-unwind` ABI
* The Rust crate containing the `-unwind` ABI declaration was compiled with
  `panic=unwind`
* The program unwinds into a Rust crate compiled with `panic=abort`
```

Reply:
```suggestion
* The program unwinds into a rust crate compiled with `panic=abort`
```
Would this be correct?

Reply:
Yes that sounds right. Cc @nbdd0121 to be sure.

Reply:
But also, don't we have to say somewhere that if a crate is built with panic=abort, and later linked with the panic=unwind runtime, that's UB too?

Reply:
In my view this is resolved by my later commits, but someone should take a look.